File: codes/FloydWarshall.cpp
/*
************************************************
username    :   smmehrab
fullname    :   s.m.mehrabul islam
email       :   mehrab.24csedu.001@gmail.com
institute   :   university of dhaka, bangladesh
session     :   2017-2018
************************************************
*/
#include<bits/stdc++.h>
using namespace std;

#define INF INT_MAX

void floydWarshall(int n, vector<vector<int>>& distances) {
	for(int k=0; k<n; k++) {
		for(int i=0; i<n; i++) {
			for(int j=0; j<n; j++) {
				if(distances[i][k] != INF && distances[k][j] != INF && distances[i][k] + distances[k][j] < distances[i][j])
					distances[i][j] = distances[i][k] + distances[k][j];
			}
		}
	}
}

int main(){
	int n, m, q, a, b, c;
    cin >> n >> m >> q;

    vector<vector<int>> distances(n, vector<int>(n, INF));
    for(int i=0; i<n; i++)
        distances[i][i] = 0;

    for(int i=0; i<m; i++) {
        cin >> a >> b >> c;
        a--; b--;
        distances[a][b] = min(distances[a][b], c);
        distances[b][a] = min(distances[b][a], c);
    }

	floydWarshall(n, distances);

	cout<<"Shortest Distances Between Every Pair of Vertices" << endl;
	for (int i=0; i<n; i++) {
		for (int j=0; j<n; j++) {
            cout << ((distances[i][j] == INF) ? "-1" : to_string(distances[i][j])) << "\t";
		}
		cout<<endl;
	}

	return 0;
}

File: codes/_algorithms/graph/floyd_warshall.py
# ************************************************
# username  :   smmehrab
# fullname  :   s.m.mehrabul islam
# email     :   smmehrabul-2017614964@cs.du.ac.bd
# institute :   university of dhaka, bangladesh
# reg       :   2017614964
# ************************************************

class Graph:

    """
        Floyd Warshall Algorithm
        O(V^3)
    """

    def __init__(self, number_of_vertices):
        self.graph = []
        self.V = number_of_vertices

    def add(self, u, v, w):
        self.graph.append([u, v, w])

    def _output(self, distances):
        print("{0}\t\t{1}".format("Vertex", "Distance"))
        for i in range(self.V):
            print("{0}\t\t{1}".format(i, distances[i]))

    def bellman_ford(self, source):

        distances = [float("Inf")] * self.V
        distances[source] = 0

        for _ in range(self.V - 1):
            for u, v, w in self.graph:
                if distances[u] != float("Inf") and distances[v] > distances[u] + w:
                    distances[v] = distances[u] + w

        for u, v, w in self.graph:
            if distances[u] != float("Inf") and distances[v] > distances[u] + w:
                print("Negative Cycle Detected!")
                return

        self._output(distances)

if __name__ == '__main__':
    graph = Graph(5)
    graph.add(0, 1, -1)
    graph.add(0, 2, 4)
    graph.add(1, 2, 3)
    graph.add(1, 3, 2)
    graph.add(1, 4, 2)
    graph.add(3, 2, 5)
    graph.add(3, 1, 1)
    graph.add(4, 3, -3)

    graph.bellman_ford(0)

